/*
*  Copyright (c) 2020, The OpenThread Authors.
*  Copyright 2024 NXP
*  All rights reserved.
*
*  Redistribution and use in source and binary forms, with or without
*  modification, are permitted provided that the following conditions are met:
*  1. Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*  2. Redistributions in binary form must reproduce the above copyright
*     notice, this list of conditions and the following disclaimer in the
*     documentation and/or other materials provided with the distribution.
*  3. Neither the name of the copyright holder nor the
*     names of its contributors may be used to endorse or promote products
*     derived from this software without specific prior written permission.
*
*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
*  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
*  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
*  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
*  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
*  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
*  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
*  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
*  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
*  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
*  POSSIBILITY OF SUCH DAMAGE.
*/

/*
* Header for patch information.
* Description           : v001 - OT host would send redundant spinel SET PHY CHANNEL command switching to
*                         CSL secondary channel much before CSL Transmission is scheduled which would
*                         hinder the current Rx on PHY Channel.
*                         v002 - Updated init value of mRxOnWhenBackoff and also when its false, BeginTransmit
*                         function does Rx on aChannel mommentarily as required by other operations
*                         [Active Scan, OT_RADIO_CAPS_SLEEP_TO_TX].
* Version               : v003
* Mandatory/Optional    : Mandatory
* Impact if not applied : Switching to CSL secondary channel few milli-seconds before CSL Transmission
*                         would result in loss of Rx packets on PHY channel degrading dual PAN Performance.
*/

diff --git a/src/core/mac/sub_mac.cpp b/src/core/mac/sub_mac.cpp
index edeba2e8b..0b6c073db 100644
--- a/src/core/mac/sub_mac.cpp
+++ b/src/core/mac/sub_mac.cpp
@@ -76,7 +76,7 @@ void SubMac::Init(void)
     mTransmitRetries = 0;
     mShortAddress    = kShortAddrInvalid;
     mExtAddress.Clear();
-    mRxOnWhenIdle      = true;
+    mRxOnWhenIdle      = false; //mRxOnWhenIdle shall be false, as its init value is false.
     mEnergyScanMaxRssi = Radio::kInvalidRssi;
     mEnergyScanEndTime = Time{0};
 #if OPENTHREAD_CONFIG_MAC_ADD_DELAY_ON_NO_ACK_ERROR_BEFORE_RETRY
@@ -567,12 +567,13 @@ void SubMac::BeginTransmit(void)
     VerifyOrExit(mState == kStateCsmaBackoff);
 #endif
 
-    if ((mRadioCaps & OT_RADIO_CAPS_SLEEP_TO_TX) == 0)
+/* - Enable RX only when mRxOnWhenIdle=0 which is one-to-one mapped with mRxOnWhenBackoff, and OT_RADIO_CAPS_SLEEP_TO_TX=0.
+ * - Don't send spinel SET CHANNEL CMD followed by starting Rx on secondary channel just before CSL Tx, which may take few mSec, and this will block the reception on the Primary channel.*/
+    if ((mRadioCaps & OT_RADIO_CAPS_SLEEP_TO_TX) == 0 && !mRxOnWhenIdle)
     {
         SuccessOrAssert(Get<Radio>().Receive(mTransmitFrame.GetChannel()));
     }
 
-    SetState(kStateTransmit);
 
     if (mPcapCallback.IsSet())
     {
